### ISSUE TITLE
Make `plasmapy.formulary` compatible with `CustomParticle` and `ParticleList`

### ISSUE THREAD

**Issue by namurphy** [PlasmaPy/PlasmaPy#1716]
Right now, `CustomParticle` and `ParticleList` are not compatible with most functions in `plasmapy.formulary`. However, it would be very powerful to be able to calculate quantities like the gyroradius or the plasma frequency for all the particles in a plasma at once, or for custom (e.g., mean or composite) particles.

Once #1057 is merged, `@particle_input` will be able to pass through or create `CustomParticle` and `ParticleList` objects from annotated arguments.  More consistent use of `@particle_input` in `plasmapy.formulary` (#341) will facilitate greater compatibility with `CustomParticle` and `ParticleList` objects.  

Some considerations...

* It will be vital to have tests that provide `CustomParticle` and `ParticleList` objects to different functions.  

* The annotations will need to change from `Particle` → `ParticleLike`.

* We'll probably want to use `|particle-like|` in the type information in the docstrings to link to the glossary entry.

* The ``allow_custom_particles`` and ``allow_particle_lists`` to `@particle_input` may need to be set to `False` in some cases (e.g., having multiple ions for `Alfven_speed` might not be sensible).

* [out of scope?] There are a few functions or methods with parameters that are annotated with ``(Particle, Particle)`` to indicate a particle pair.  This could be replaced with `ParticleListLike`...and we might even be able to do things like `ParticleListLike[2]` to indicate that it should have two items (though that might need Python 3.9+ and/or `typing_extensions`.  It will probably be sensible to leave these for later, after some more development of `@particle_input`.
